File: 221_maximal_square_bfs.cpp
#include <vector>
#include <tuple>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int M, N;
    vector<vector<char>> matrix;

    vector<pair<int, int>> adjs(int row, int col)
    {
        vector<pair<int, int>> neighbors;
        vector<pair<int, int>> candidates = {{row + 1, col}, {row, col + 1}, {row + 1, col + 1}};

        for (auto [a, b] : candidates)
        {
            if (0 <= a && a < M && 0 <= b && b < N)
            {
                neighbors.emplace_back(a, b);
            }
        }
        return neighbors;
    }

    vector<pair<int, int>> expand_sqr(const vector<pair<int, int>> &sqr)
    {
        vector<pair<int, int>> next_sqr;

        for (const auto &[p, q] : sqr)
        {
            for (const auto &[a, b] : adjs(p, q))
            {
                // Check if (a, b) is already in sqr or next_sqr
                if (find(sqr.begin(), sqr.end(), make_pair(a, b)) == sqr.end() &&
                    find(next_sqr.begin(), next_sqr.end(), make_pair(a, b)) == next_sqr.end())
                {
                    if (matrix[a][b] == '1')
                    {
                        next_sqr.emplace_back(a, b);
                    }
                    else
                    {
                        return {};
                    }
                }
            }
        }

        return (next_sqr.size() == sqr.size() + 2) ? next_sqr : vector<pair<int, int>>{};
    }

    int bfs(int i, int j)
    {
        vector<pair<int, int>> sqr = {{i, j}};
        int n = 0;

        while (true)
        {
            vector<pair<int, int>> next_sqr = expand_sqr(sqr);
            n += 1;

            if (!next_sqr.empty())
            {
                sqr = next_sqr;
            }
            else
            {
                break;
            }
        }
        return n;
    }

    int maximalSquare(vector<vector<char>> &matrix)
    {
        this->matrix = matrix;
        M = matrix.size();
        if (M == 0)
            return 0;
        N = matrix[0].size();
        int maximal = 0;

        for (int i = 0; i < M; ++i)
        {
            for (int j = 0; j < N; ++j)
            {
                if (matrix[i][j] == '1')
                {
                    int cr = i + 1;
                    while (cr < M && matrix[cr][j] == '1')
                        ++cr;

                    int cc = j + 1;
                    while (cc < N && matrix[i][cc] == '1')
                        ++cc;

                    if ((min(cr - i, cc - j)) * (min(cr - i, cc - j)) <= maximal)
                        continue;

                    int n = bfs(i, j);
                    maximal = max(maximal, n * n);
                }
            }
        }
        return maximal;
    }
};

File: 221_maximal_square_bfs.py
from typing import List


# This solution exceeds the time limit for the problem
class Solution:
    def adjs(self, row: int, col: int):
        return (
            (a, b)
            for a, b in ((row + 1, col), (row, col + 1), (row + 1, col + 1))
            if 0 <= a < self.M and 0 <= b < self.N
        )

    def expand_sqr(self, sqr) -> list[tuple[int, int]] | None:
        next_sqr = []

        for p, q in sqr:
            for a, b in self.adjs(p, q):
                if (a, b) not in sqr and (a, b) not in next_sqr:
                    if self.matrix[a][b] == "1":
                        next_sqr.append((a, b))
                    else:
                        return None

        return next_sqr if len(next_sqr) == len(sqr) + 2 else None

    def bfs(self, i: int, j: int) -> int:
        sqr = [(i, j)]
        n = 0

        while True:
            next_sqr = self.expand_sqr(sqr)
            n += 1

            if next_sqr:
                sqr = next_sqr
            else:
                break

        return n

    def maximalSquare(self, matrix: List[List[str]]) -> int:
        self.M, self.N = len(matrix), len(matrix[0])
        self.matrix = matrix
        maximal = 0

        for i in range(self.M):
            for j in range(self.N):
                if self.matrix[i][j] == "1":
                    cr = i + 1

                    while cr < self.M and self.matrix[cr][j] == "1":
                        cr += 1

                    cc = j + 1

                    while cc < self.N and self.matrix[i][cc] == "1":
                        cc += 1

                    if min(cr - i, cc - j) ** 2 <= maximal:
                        continue

                    n = self.bfs(i, j)
                    maximal = max(maximal, n**2)
        return maximal


if __name__ == "__main__":
    s = Solution()

    matrix = [
        ["1", "0", "1", "0", "0"],
        ["1", "0", "1", "1", "1"],
        ["1", "1", "1", "1", "1"],
        ["1", "0", "0", "1", "0"],
    ]

    ans = s.maximalSquare(matrix)
    print(ans, "expected 4")

    matrix = [["0", "1"], ["1", "0"]]

    ans = s.maximalSquare(matrix)
    print(ans, "expected 1")

    matrix = [["0"]]

    ans = s.maximalSquare(matrix)
    print(ans, "expected 0")

    matrix = [
        ["1", "1", "1"],
        ["1", "1", "1"],
        ["1", "1", "1"],
    ]

    ans = s.maximalSquare(matrix)
    print(ans, "expected 9")

    matrix = [
        ["1", "1", "1", "1", "0"],
        ["1", "1", "1", "1", "0"],
        ["1", "1", "1", "1", "1"],
        ["1", "1", "1", "1", "1"],
        ["0", "0", "1", "1", "1"],
    ]

    ans = s.maximalSquare(matrix)
    print(ans, "expected 16")
